modules/imgproc/src/filter.cpp

```diff
@@ -4610,24 +4610,48 @@ static bool ipp_filter2D( InputArray _src, OutputArray _dst, int ddepth,
 
             if ((status = ippiFilterBorderGetSize(kernelSize, dstRoiSize, dataType, kernelType, cn, &specSize, &bufsize)) >= 0)
             {
-                IppiFilterBorderSpec * spec = (IppiFilterBorderSpec *)ippMalloc(specSize);
-                Ipp8u * buffer = ippsMalloc_8u(bufsize);
+                IppAutoBuffer<IppiFilterBorderSpec> spec(specSize);
+                IppAutoBuffer<Ipp8u> buffer(bufsize);
                 Ipp32f borderValue[4] = { 0, 0, 0, 0 };
 
-                Mat reversedKernel;
-                flip(kernel, reversedKernel, -1);
-
-                if ((kdepth == CV_32F && (status = ippiFilterBorderInit_32f((const Ipp32f *)reversedKernel.data, kernelSize,
-                    dataType, cn, ippRndFinancial, spec)) >= 0 ) ||
-                    (kdepth == CV_16S && (status = ippiFilterBorderInit_16s((const Ipp16s *)reversedKernel.data,
-                    kernelSize, 0, dataType, cn, ippRndFinancial, spec)) >= 0))
+                if(kdepth == CV_32F)
                 {
-                    status = ippFunc(src.data, (int)src.step, dst.data, (int)dst.step, dstRoiSize,
-                        ippBorderType, borderValue, spec, buffer);
+                    Ipp32f *pKerBuffer = (Ipp32f*)kernel.data;
+                    IppAutoBuffer<Ipp32f> kerTmp;
+                    if(kernel.step != kernel.cols)
```

Comment:
Missed element size: `kernel.cols * sizeof(float)` here and similar fix below.

P.S. `CV_ELEM_SIZE1(kdepth)` should also work instead of sizeof(float)